clang/lib/CIR/CodeGen/CIRGenBuiltin.cpp

```diff
@@ -137,6 +137,96 @@ EncompassingIntegerType(ArrayRef<struct WidthAndSignedness> Types) {
   return {Width, Signed};
 }
 
+/// Emit the conversions required to turn the given value into an
+/// integer of the given size.
+static mlir::Value buildToInt(CIRGenFunction &CGF, mlir::Value v, QualType t,
+                              mlir::cir::IntType intType) {
+  v = CGF.buildToMemory(v, t);
+
+  if (isa<mlir::cir::PointerType>(v.getType()))
+    return CGF.getBuilder().createPtrToInt(v, intType);
+
+  assert(v.getType() == intType);
+  return v;
+}
+
+static mlir::Value buildFromInt(CIRGenFunction &CGF, mlir::Value v, QualType t,
+                                mlir::Type resultType) {
+  v = CGF.buildFromMemory(v, t);
+
+  if (isa<mlir::cir::PointerType>(resultType))
+    return CGF.getBuilder().createIntToPtr(v, resultType);
+
+  assert(v.getType() == resultType);
+  return v;
+}
+
+static Address checkAtomicAlignment(CIRGenFunction &CGF, const CallExpr *E) {
+  ASTContext &ctx = CGF.getContext();
+  Address ptr = CGF.buildPointerWithAlignment(E->getArg(0));
+  unsigned bytes =
+      isa<mlir::cir::PointerType>(ptr.getElementType())
+          ? ctx.getTypeSizeInChars(ctx.VoidPtrTy).getQuantity()
+          : CGF.CGM.getDataLayout().getTypeSizeInBits(ptr.getElementType()) / 8;
+  unsigned align = ptr.getAlignment().getQuantity();
+  if (align % bytes != 0) {
+    DiagnosticsEngine &diags = CGF.CGM.getDiags();
+    // TODO
+    // diags.Report(E->getBeginLoc(), diag::warn_sync_op_misaligned);
+
+    // Force address to be at least naturally-aligned.
+    return ptr.withAlignment(CharUnits::fromQuantity(bytes));
+  }
+  return ptr;
+}
+
+/// Utility to insert an atomic instruction based on Intrinsic::ID
+/// and the expression node.
+static mlir::Value
+makeBinaryAtomicValue(CIRGenFunction &cgf, mlir::cir::AtomicFetchKind kind,
+                      const CallExpr *expr,
+                      mlir::cir::MemOrder ordering =
+                          mlir::cir::MemOrder::SequentiallyConsistent) {
+
+  QualType typ = expr->getType();
+
+  assert(expr->getArg(0)->getType()->isPointerType());
+  assert(cgf.getContext().hasSameUnqualifiedType(
+      typ, expr->getArg(0)->getType()->getPointeeType()));
+  assert(
+      cgf.getContext().hasSameUnqualifiedType(typ, expr->getArg(1)->getType()));
+
+  Address destAddr = checkAtomicAlignment(cgf, expr);
+  auto &builder = cgf.getBuilder();
+  auto *ctxt = builder.getContext();
+  auto intType = builder.getSIntNTy(cgf.getContext().getTypeSize(typ));
+
+  mlir::Value val = cgf.buildScalarExpr(expr->getArg(1));
+  mlir::Type valueType = val.getType();
+  val = buildToInt(cgf, val, typ, intType);
+
+  auto op = mlir::cir::AtomicFetch::getOperationName();
+  SmallVector<mlir::Value> atomicOperands = {destAddr.emitRawPointer(), val};
+  auto fetchAttr =
+      mlir::cir::AtomicFetchKindAttr::get(builder.getContext(), kind);
+  auto rmwi = builder.create(cgf.getLoc(expr->getSourceRange()),
```

Comment:
Please create this in the canonical way `builder.create<mlir::cir::AtomicFetch>...`, we only do this approach in the place you copied this from because we need it to be more generic and cover more intrinsics.